app/server/lib/scim/v2/ScimUserController.ts

```diff
@@ -0,0 +1,121 @@
+import { ApiError } from 'app/common/ApiError';
+import { HomeDBManager, Scope } from 'app/gen-server/lib/homedb/HomeDBManager';
+import SCIMMY from 'scimmy';
+import { toSCIMMYUser, toUserProfile } from './ScimUserUtils';
+import { RequestContext } from './ScimTypes';
+
+class ScimUserController {
+  private static _getIdFromResource(resource: any) {
+    const id = parseInt(resource.id, 10);
+    if (Number.isNaN(id)) {
+      throw new SCIMMY.Types.Error(400, 'invalidValue', 'Invalid passed user ID');
+    }
+    return id;
+  }
+
+  constructor(
+    private _dbManager: HomeDBManager,
+    private _checkAccess: (context: RequestContext) => void
+  ) {}
+
+  public async getSingleUser(resource: any, context: RequestContext) {
+    this._checkAccess(context);
+
+    const id = ScimUserController._getIdFromResource(resource);
+    const user = await this._dbManager.getUser(id);
+    if (!user) {
+      throw new SCIMMY.Types.Error(404, null!, `User with ID ${id} not found`);
+    }
+    return toSCIMMYUser(user);
+  }
+
+  public async getUsers(resource: any, context: RequestContext) {
+    this._checkAccess(context);
+
+    const { filter } = resource;
+    const scimmyUsers = (await this._dbManager.getUsers()).map(user => toSCIMMYUser(user));
+    return filter ? filter.match(scimmyUsers) : scimmyUsers;
+  }
+
+  public async createUser(data: any, context: RequestContext) {
+    this._checkAccess(context);
+
+    try {
+      await this._checkEmailIsUnique(data.userName);
+      const userProfile = toUserProfile(data);
+      const newUser = await this._dbManager.getUserByLoginWithRetry(userProfile.email, {
+        profile: userProfile
+      });
+      return toSCIMMYUser(newUser);
+    } catch (ex) {
+      return this._toScimError(ex);
+    }
+  }
+
+  public async overrideUser(resource: any, data: any, context: RequestContext) {
+    this._checkAccess(context);
+
+    try {
+      const id = ScimUserController._getIdFromResource(resource);
+      await this._checkEmailIsUnique(data.userName, id);
+      const updatedUser = await this._dbManager.overrideUser(id, toUserProfile(data));
+      return toSCIMMYUser(updatedUser);
+    } catch (ex) {
+      return this._toScimError(ex);
+    }
+  }
+
+  public async deleteUser(resource: any, context: RequestContext) {
+    this._checkAccess(context);
+
+    const id = ScimUserController._getIdFromResource(resource);
+    try {
+      const fakeScope: Scope = { userId: id };
+      // FIXME: deleteUser should probably better not requiring a scope.
+      await this._dbManager.deleteUser(fakeScope, id);
```

Comment:
It feels like a bit dirty to me to create a fake Scope, but I fear it is hard to get rid of it from `UsersManager.deleteUser`, especially because the [`deleteOrg`](https://github.com/gristlabs/grist-core/blob/243369513f220aa0b576c884b02aea0554539b1d/app/gen-server/lib/homedb/UsersManager.ts#L496) needs it.

Any opinion/idea on that?

Reply:
My concern here is that the scope is only required to enforce that users can only delete themselves. I suppose since SCIM is only exposed to one or two admin users, it's okay to allow a fake context.